app/api/src/players/players.service.ts

```diff
@@ -42,4 +43,129 @@ export class PlayersService {
 		}
 		return player;
 	}
+
+	async findTop(topPlayersQuery: TopPlayersQueryDto) {
+		const { stat: hitsStat, gameType, minGames, minShots, limit } = topPlayersQuery;
+
+		const shotsStat = {
+			discDmgHitsTG: 'discShotsFiredTG',
+			discHitsTG: 'discShotsFiredTG',
+			discMATG: 'discShotsFiredTG',
+			laserHitsTG: 'laserShotsFiredTG',
+			laserMATG: 'laserShotsFiredTG',
+			cgHitsTG: 'cgShotsFiredTG',
+			shockHitsTG: 'shockShotsFiredTG'
+		}[hitsStat]
+
+		// Possibly make this a query param at some point?
+		const excludeDiscJumps = shotsStat === 'discShotsFiredTG';
+
+		const hitsValue = '(game.stats->:hitsStat->>0)::integer';
+		const shotsValue = '(game.stats->:shotsStat->>0)::integer';
+		const discJumpsValue = "(game.stats->'discJumpTG'->>0)::integer";
+		const killerDiscJumpsValue = "(game.stats->'killerDiscJumpTG'->>0)::integer";
+
+		// pgSQL doesn't let you reference aliased selections you've made in the
+		// same select statement, so unfortunately these computed JSON values are
+		// repeated in a couple places. Using template strings here to easily repeat
+		// the same expression instead of having multiple nested subqueries. Rest
+		// assured, no user-supplied values are ever interpolated, those are all
+		// parameterized instead.
+		const aggregatedHits = `SUM(${hitsValue})::integer`;
+		// Add sums of `discJumpTG` and `killerDiscJumpTG`.
+		const aggregatedDiscJumps =
+			`(SUM(${discJumpsValue})::integer + SUM(${killerDiscJumpsValue})::integer)`;
+
+		let aggregatedShots = `SUM(${shotsValue})::integer`;
+		if (excludeDiscJumps) {
+			// Since subtracting disc jumps could theoretically drop the shots count
+			// to 0, clamp it to at least the number of hits, otherwise it'd be
+			// possible to have >100% accuracy.
+			aggregatedShots = `GREATEST(${aggregatedHits}, ${aggregatedShots} - ${aggregatedDiscJumps})`
+		}
+
+		// Cast to float to avoid integer division truncating the result.
+		const aggregatedAccuracy = `(${aggregatedHits}::float / ${aggregatedShots}::float)`;
+
+		// TODO: This whole query could probably be turned into a `ViewEntity` at
+		// some point, but I couldn't get that to work.
```

Comment:
I tried making a ViewEntity, but no luck. I think the only issue was it synchronizing with the database to actually do the `CREATE VIEW`, but even trying `synchronize: true` didn't help. Didn't have the patience to get it working for now.

Reply:
Also, having it as a view might limit the flexibility in the query anyway (since a bunch of options are passed as parameters that affect the query, e.g. `gameType`, `excludeDiscJumps`).